### PR TITLE
(knative, kfserving, cert-manager) Build working solution for kfserving integration

### PR DESCRIPTION
**Pull request by zijianjoy**
Fix: https://github.com/kubeflow/gcp-blueprints/issues/209
Fix: #222
Fix: https://github.com/kubeflow/gcp-blueprints/issues/176

Partial: #221

Using the following versions:

KFserving 0.5.1
KNative 0.22.0
Cert-manager 1.3

This is the only combination which is validated to be working for KFServing integration.

KFServing team recommended KNative 0.17.4, but the `kubeflow/manifests` didn't update the manifest for this version, plus that   KNative official doc has already archived this version. I tried to install this version but installation doesn't work for inferenceService deployment.

David has a PR for cert-manager upgrade https://github.com/kubeflow/manifests/pull/1820, but it needs some adjustment for successful deployment.

Tried to adopt KNative installation in kustomize build rule but kustomize complains about existing registered ID for serving-core.yaml file. Currently we use command line for installing kNative.

### CODE REVIEW

**Review by Bobgy — COMMENTED** [review]
/lgtm
/approve

Looks good to me!
You can decide when to unhold
/hold

**Comment by DavidSpek** [comment]
I've also updated the Argo CD installation to knative 0.22.0. The problem with the existing registered ID, is that the `knative-core.yaml` contains all the entries of the `knative-crds.yaml`. In the following commit you can see a working kustomize folder for knative 0.22.0. https://github.com/argoflow/argoflow/commit/ce115f089736b8e2988fa31029eb1d3411fc1be4

Regarding cert-manager, I'm not sure what trouble you were experiencing with the installation. But I'm around on Slack to debug this as it worked fine with plane kustomize for me.

File: kubeflow/Makefile [hunk 1]
```diff
@@ -189,6 +179,45 @@ apply-kubeflow: validate-values check-name
 	# https://github.com/kubeflow/gcp-blueprints/issues/160
 	kubectl --context=$(KFCTXT) -n istio-system delete pods -l service=backend-updater
 
+.PHONY: apply-knative
+apply-knative:
+	# Common Knative
+	# It has error when trying to patch the resource downloaded from knative because it doesn't match kustomize requirement.
+	# Original release of knative manifest can not be patched by kustomize because of `already registered id`. For now we run kubectl directly.
```
[the remark this answers is not in the document]
Comment (DavidSpek) [reply kubeflow/Makefile:186]: There are 2 issues that I came across with regards to the knative manifests. The first is a bug in kustomize causing YAML anchors to not work properly. The second is that the `serving-core.yaml` contains all of the CRDs that are in the `serving-crds.yaml` file. As such, you can't use both when using Kustomize, but you might run into a chicken & egg so you need to apply the manifests twice if your not using something that continuously syncs like Argo CD.

I expanded the manifests to remove the YAML anchors and created the following kustomize folder for the Argo CD installation: https://github.com/argoflow/argoflow/commit/ce115f089736b8e2988fa31029eb1d3411fc1be4.

[the remark this answers is not in the document]
Comment (DavidSpek) [reply kubeflow/Makefile:186]: I believe a decision was made to not update KNative (or cert-manager for that matter) in the manifests repo before the 1.3 release. So I believe the route forward is to add these manifests in the gcp-blueprints repo directly for now, and try and get the fixes upstreamed after the release.

[the remark this answers is not in the document]
Comment (zijianjoy) [reply kubeflow/Makefile:183]: Thank you for clarifying, let's merge the fix in `kubeflow/manifests` after the release. We will do the patch for now on gcp-blueprints.

**Comment by DavidSpek** [comment]
@zijianjoy Thought I’d link to my comment about the problem you were having with my cert-manager PR in case you missed it. https://github.com/kubeflow/manifests/pull/1820#issuecomment-822528568

**Comment by connorlwilkes** [comment]
Apologies but just reading this through. Why is there a difference in knative version here vs in manifests repository?

**Comment by zijianjoy** [comment]
> Just reading this through and was wondering why is there a difference in knative version here vs in manifests repository?

@connorlwilkes Upstream manifest has KNative v0.14.3: https://github.com/kubeflow/manifests/blob/master/common/knative/knative-serving-crds/base/crd.yaml#L7. But KFserving requires KNative v0.17.4 or above: https://github.com/kubeflow/kfserving#prerequisites

**Comment by zijianjoy** [comment]
> @zijianjoy Thought I’d link to my comment about the problem you were having with my cert-manager PR in case you missed it. [kubeflow/manifests#1820 (comment)](https://github.com/kubeflow/manifests/pull/1820#issuecomment-822528568)

@DavidSpek Thank you for the link! I make the changes below to the cert-manager manifest. It allows me to successfully deploy cert-manager and kubeflow-issuer:

1. Move cluster-issuer.yaml to `cert-manager/kubeflow-issuer/`
2. Update `cert-manager/kubeflow-issuer/kustomization.yaml` to add `resources` cluster-issuer.yaml 
3. Update `cert-manager/overlays/self-signed/kustomization.yaml` to use `resources` cert-manager/kubeflow-issue instead of cluster-issuer.yaml.
2. More detail in this PR commit: kubeflow/common/cert-manager/cert-manager-1-3

When applying these resource, I apply the cert-manager/base first, then apply the cert-manager/kubeflow-issuer. It avoids kustomize to apply the following labels to all cert-manager deployments:
```

commonLabels:
  kustomize.component: cert-manager
  app.kubernetes.io/component: cert-manager
  app.kubernetes.io/name: cert-manager
```

Otherwise, cert-manager service on cluster will be mapped to 3 deployments: cert-manager, cert-manager-cainjector, cert-manager-webhook, which is not right. cert-manager service should point to cert-manager deployment only. But `overlays/self-signed` enforced all these labels on base resource.

**Review by Bobgy — COMMENTED** [review]
Great job!
/lgtm
/approve

**Comment by zijianjoy** [comment]
Thank you so much Yuan for your help!

/unhold

**Pull request merged**